Bug 1530359 - Handle URL redirects for CFR triggers

piatra opened the pull request:
The way I'm handling redirects is to additionally check the original request host against the hosts defined in the message.
For example if `de.quora.com` fails the initial check but we initially navigated to `quora.com` then the check will now pass.
This doesn't mean that `bar.mozilla.com` will match for `mozilla.com` because we check that `(originalLocation !== location)` before trying the second lookup so it's only for redirects when this check would return true.

k88hudson commented:
Another thing you could consider is adding a ` onStateChange` listener and checking `aStateFlags & Ci.nsIWebProgressListener.STATE_REDIRECTING` / `aRequest.originalURL`, although I'm not sure what's better v.s. just looking at `aRequest.QueryInterface(Ci.nsIChannel).originalURI` every time. I'd go with whatever is the least expensive if there's a meaningful difference.

Also this is kinda unrelated but I happened to notice – is there a reason we need to create a new `URL` object instead of just using `aLocationURI.host` (and `originalLocation.host` for that matter)?

piatra commented:
I added a `onStateChange` handler and attempted the same thing. Some things I noticed:
* `onStateChange` is fired after `onLocationChange`
  * this was surprising, I expected the other way around and planned on using `onStateChange` to set some state that we could later use in `onLocationChange`
* `onStateChange` also fires for domains that redirect to and from the `www.` version of the website
  * `foo.com` redirects to `www.foo.com`
  * we would risk calling the `triggerHandler` twice in this scenario
* `aStateFlags & Ci.nsIWebProgressListener.STATE_REDIRECTING` was never `true` for `quora.com` even though the request got redirected
  * `aRequest.originalURL` was different from `aRequest.URI` and this is what I used to confirm the redirect attempt

The current implementation that checks for original vs current URL doesn't seem to have a significant cost to it. I also updated `checkHost` to not create the extra URL objects and reuse the `aLocation` arguments (the profiler screenshot was taken before)
<img width="413" alt="image" src="https://user-images.githubusercontent.com/810040/54683803-82ef7c80-4b0a-11e9-89db-9739f26828fe.png">

k88hudson commented on lib/ASRouterTriggerListeners.jsm:
```diff
@@ -33,6 +33,29 @@ function isPrivateWindow(win) {
+ * Additionally verify for redirects and check original request URL against
+ * the whitelist.
+ *
+ * @returns {string} - the host of the URL that matched
```
This should be `@returns {boolean}`

k88hudson commented on lib/ASRouterTriggerListeners.jsm:
```diff
@@ -33,6 +33,29 @@ function isPrivateWindow(win) {
+ *
+ * @returns {string} - the host of the URL that matched
+ */
+function checkHost(location, hosts, aRequest) {
```
maybe name this `aLocationURI` for consistency?

k88hudson commented on lib/ASRouterTriggerListeners.jsm:
```diff
@@ -252,18 +272,15 @@ this.ASRouterTriggerListeners = new Map([
-            this._triggerHandler(aBrowser, {id: "openURL", param: host});
-          }
-        } catch (e) {} // Couldn't parse location URL
+      if (host && aWebProgress.isTopLevel && !isSameDocument) {
```
You could combine these if statements if you want

piatra commented on lib/ASRouterTriggerListeners.jsm:
```diff
@@ -33,6 +33,29 @@ function isPrivateWindow(win) {
+ * Additionally verify for redirects and check original request URL against
+ * the whitelist.
+ *
+ * @returns {string} - the host that matched the whitelist
```
I went back on this and now returning the `host` that matched the whitelist because we have an additional check 
https://github.com/mozilla/activity-stream/blob/b89bae21fd28625f3d70cadb4f1a404340aeb129/lib/ASRouterTargeting.jsm#L341
Where ensure that the selected message matches the trigger. If we would have passed down `aLocationURI.host` that could be `redirect.foo.com` and not match `foo.com` even if `checkHost` initially returned `true`.

piatra commented on lib/CFRMessageProvider.jsm:
```diff
@@ -38,12 +38,10 @@ const REDDIT_ENHANCEMENT_PARAMS = {
-  "reddit.com", "www.reddit.com", "github.com", "www.github.com", "youtube.com", "www.youtube.com",
-  "feedly.com", "www.feedly.com", "drive.google.com", "amazon.com", "www.amazon.com",
-  "messages.android.com",
+  "docs.google.com", "calendar.google.com",
```
Because we're only comparing hosts now I don't think we need the `www.` version anymore. Any thoughts on this?

k88hudson replied:
`www.google.com` is a different host than `google.com` so I think you still need both (or, functionally speaking with this patch, whatever the redirection ends up at)

k88hudson commented on test/unit/asrouter/ASRouterTriggerListeners.test.js:
```diff
@@ -178,10 +178,59 @@ describe("ASRouterTriggerListeners", () => {
+        };
+        openURLListener.onLocationChange(browser, webProgress, aRequest, aLocationURI);
+        assert.calledOnce(aRequest.QueryInterface);
+        assert.notCalled(newTriggerHandler);
```
Might be worth adding a `www` test specifically since someone else might forget that's a subdomain?

k88hudson replied:
or maybe just add a comment near the actual list

piatra commented:
I added a test to make sure we don't try to remove it again.

k88hudson commented on test/unit/asrouter/CFRMessageProvider.test.js:
```diff
@@ -41,4 +41,9 @@ describe("CFRMessageProvider", () => {
       }
     }
   });
+  it("should contain `www.` version of the hosts", () => {
```
:+1:

k88hudson commented on lib/ASRouterTriggerListeners.jsm:
```diff
@@ -152,7 +152,12 @@ this.ASRouterTriggerListeners = new Map([
+      let host;
+      try {
+        host = aLocationURI ? aLocationURI.host : "";
+      } catch (e) { // about: pages will throw errors
```
Ah, good call.
